### PR TITLE
add cloudforms inventory script

### PR DESCRIPTION
**Pull request by mrjoshuap**
[cloudforms--list.json.txt](https://github.com/ansible/ansible/files/412082/cloudforms--list.json.txt)
[cloudforms--host.json.txt](https://github.com/ansible/ansible/files/412086/cloudforms--host.json.txt)
##### ISSUE TYPE
- Feature Pull Request
##### COMPONENT NAME

Contributed inventory for Red Hat CloudForms / ManageIQ
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### SUMMARY

This is a new contributed inventory script for Red Hat CloudForms / ManageIQ.  I tested with Red Hat CloudForms 4.1.  See attached output examples.

```
usage: cloudforms.py [-h] [--list] [--host HOST] [--refresh-cache] [--debug]

Produce an Ansible Inventory file based on CloudForms managed VMs

optional arguments:
  -h, --help       show this help message and exit
  --list           List instances (default: True)
  --host HOST      Get all the variables about a specific instance
  --refresh-cache  Force refresh of cache by making API requests to cloudforms
                   (default: False - use cache files)
  --debug          Show debug output while running (default: False)
```

based on the foreman inventory script, features:
- cached results (default 600 seconds)
- paginated host results (default 100 hosts)
- ssl verification (default True)
- arguments to flush cache and run in debug mode

### CODE REVIEW

File: contrib/inventory/cloudforms.py [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+        if self.args.host:
+            if self.debug:
+                print "Fetching host [%s]" % self.args.host
+            data_to_print += self.get_host_info(self.args.host)
+        else:
+            self.inventory['_meta'] = {'hostvars': {}}
+            for hostname in self.cache:
+                self.inventory['_meta']['hostvars'][hostname] = {
+                    'cloudforms': self.cache[hostname],
+                    'cloudforms_params': self.params[hostname],
```
Comment (nitzmahone) [inline contrib/inventory/cloudforms.py]: Why the two dicts? Seems like some stuff is duplicated between cloudforms and cloudforms_params- would it be more user-friendly to squash to a single "cloudforms_facts" dict or something?

**Comment by nitzmahone** [comment]
Sweet- just the one question about the two sub-dicts, otherwise LGTM!

**Comment by mrjoshuap** [comment]
There is no technical reason they can't be squashed into a single dict; this was based on the foreman.py inventory script that shipped with Ansible Tower 3.0.1, which used the second for host group parameters.  Let me rework it with the suggestion and update this PR.

**Comment by mrjoshuap** [comment]
Reworked to squash into a single dict.  Additionally:
- removed second cache / dict with duplicate info
- added purge_actions configuration option to remove the actions from a host (defaults to False)
- added prefer_ip_address configuration option so give the option of using ip address instead of name (defaults to True)
- removed self variables — just use the arguments directly
- added --pretty command line option to pretty print results
- renamed _resolve_params to _resolve_host

New output attached.

[cloudforms--host.json.txt](https://github.com/ansible/ansible/files/415494/cloudforms--host.json.txt)
[cloudforms--list.json.txt](https://github.com/ansible/ansible/files/415495/cloudforms--list.json.txt)

**Comment by mrjoshuap** [comment]
Also, --help updated:

```
usage: cloudforms.py [-h] [--list] [--host HOST] [--pretty] [--refresh-cache]
                     [--debug]

Produce an Ansible Inventory file based on CloudForms managed VMs

optional arguments:
  -h, --help       show this help message and exit
  --list           List instances (default: True)
  --host HOST      Get all the variables about a specific instance
  --pretty         Pretty print JSON output (default: False)
  --refresh-cache  Force refresh of cache by making API requests to CloudForms
                   (default: False - use cache files)
  --debug          Show debug output while running (default: False)
```

File: contrib/inventory/cloudforms.py [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+            if host.has_key('tags'):
+                for group in host['tags']:
+                    safe_key = self.to_safe(group['name'])
+                    self.push(self.inventory, safe_key, dns_name)
+
+                    if self.args.debug:
+                        print "Found tag [%s] for host which will be mapped to [%s]" % (group['name'], safe_key)
+
+            full_host = self._resolve_host(host)
+            if host.has_key('ipaddresses'):
```
Comment (jlaska) [inline contrib/inventory/cloudforms.py]: Can you use `if 'ipaddresses' in host` instead?  We should use this where ever you have `has_key()`

File: contrib/inventory/cloudforms.py [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+            if self.cloudforms_prefer_ip_address and host.has_key('ipaddresses') and host['ipaddresses'][0]:
+                dns_name = host['ipaddresses'][0]
+            else:
+                dns_name = host['name']
+
+            if self.args.debug:
+                print "Host named [%s] will use [%s] as dns_name" % (host['name'], dns_name)
+
+            # Create ansible groups for hostgroup, location and organization
+            if host.has_key('tags'):
```
Comment (jlaska) [inline contrib/inventory/cloudforms.py]: Suggest changing to `if 'tags' in host:`

File: contrib/inventory/cloudforms.py (hunk 2 above)
Comment (mrjoshuap) [inline contrib/inventory/cloudforms.py]: I'll rework both of them as suggested and catch any others.

File: contrib/inventory/cloudforms.py [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+        if self.args.debug:
+            print "Updating cache..."
+
+        for host in self._get_hosts():
+            if host['power_state'] != 'on':
+                if self.args.debug:
+                    print "Skipping %s because power_state = %s" % (host['name'], host['power_state'])
+                continue
+
+            if self.cloudforms_prefer_ip_address and host.has_key('ipaddresses') and host['ipaddresses'][0]:
```
Comment (jlaska) [inline contrib/inventory/cloudforms.py]: Instead of changing the `name`, can we set `ansible_ssh_host` to the discovered `ipaddress`?

Comment (jlaska) [inline contrib/inventory/cloudforms.py]: Also, this results in a traceback for me.  I think you want to do ...

``` python
if 'ipaddresses' in host and isinstance(host['ipaddresses'], list) and host['ipaddresses']:
```

**Comment by mrjoshuap** [comment]
I'm addressing all of those concerns and adapting it to catch more exceptions.

**Comment by mrjoshuap** [comment]
All issues and suggestions addressed, needs reviewed and tested by others :-)

```
usage: cloudforms.py [-h] [--list] [--host HOST] [--pretty] [--refresh-cache]
                     [--debug]

Produce an Ansible Inventory file based on CloudForms managed VMs

optional arguments:
  -h, --help       show this help message and exit
  --list           List instances (default: True)
  --host HOST      Get all the variables about a specific instance
  --pretty         Pretty print JSON output (default: False)
  --refresh-cache  Force refresh of cache by making API requests to CloudForms
                   (default: False - use cache files)
  --debug          Show debug output while running (default: False)
```

Also updated outputs:

[cloudforms_py--host.json.txt](https://github.com/ansible/ansible/files/423716/cloudforms_py--host.json.txt)
[cloudforms_py--list.json.txt](https://github.com/ansible/ansible/files/423717/cloudforms_py--list.json.txt)

File: contrib/inventory/cloudforms.py (hunk 3 above)
Comment (mrjoshuap) [inline contrib/inventory/cloudforms.py]: done 👍

File: contrib/inventory/cloudforms.py (hunk 4 above)
Comment (mrjoshuap) [inline contrib/inventory/cloudforms.py]: this is actually not needed since it will keep the name and specify `ansible_ssh_host` as the first IP.

**Comment by mrjoshuap** [comment]
Sidebar:

Can anyone confirm if group names need to be sanitized?  I have checked and searched but not found anything definitive.  For example, one of the tags returned is named **/managed/folder_path_yellow/datacenters** and a type is returned as **ManageIQ::Providers::Redhat::InfraManager::Vm** ... I believe slashes and colons will prevent Ansible from using them as groups or worse?

I added a configuration option to make it optional and defaulted it to True because I wasn't sure about the character validity in Ansible...

They are now added (as groups) as:
- _managed_folder_path_yellow_datacenters
- ManageIQ__Providers__Redhat__InfraManager__Vm

I could argue that at least for tags that I might want to make group -> subgroup split on the slashes... others are not as straight forward.

**Comment by mrjoshuap** [comment]
It is not a good idea to split tags on the slashes (/) ... the last sub-group key can be duplicated in multiple parents squashing all hosts into a single group and wreaking havoc upon unsuspecting innocents.

**Comment by mrjoshuap** [comment]
I don't have enough data to confirm my previous comment re: splitting tags on /... I created a new branch where this could be tested but not act as a blocker for this PR.

https://github.com/mrjoshuap/ansible/tree/cloudforms_inventory_nested_tags

**Comment by jlaska** [comment]
@mrjoshuap in your sample JSON output, you incude the `actions` key for each host.  Not a show stopper, but I wonder if that data is useful to ansible?  Should it be excluded in the hostvars?

File: contrib/inventory/cloudforms.py [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+        ret = requests.get(url,
+                           auth=HTTPBasicAuth(self.cloudforms_username, self.cloudforms_pw),
+                           verify=self.cloudforms_ssl_verify)
+
+        ret.raise_for_status()
+
+        try:
+            results = json.loads(ret.text)
+        except ValueError:
+            warnings.warn("Unexpected response from {0} ({1}): {2}".format(self.cloudforms_url, r.status_code, r.reason))
```
Comment (jlaska) [inline contrib/inventory/cloudforms.py]: heads up, `r` is undefined.

**Comment by jlaska** [comment]
Nit pick, but running this through `flake8` will highlight a few other minor gotchas.

File: contrib/inventory/cloudforms.py (hunk 5 above)
Comment (mrjoshuap) [inline contrib/inventory/cloudforms.py]: Whoops!  silly me.  On that now.

**Comment by mrjoshuap** [comment]
I agree with regards to actions; I made it configurable and default it to True; should I default it to False?  Also, I will utilize the flake8 and see what it says.

File: contrib/inventory/cloudforms.py (hunk 5 above)
Comment (jlaska) [inline contrib/inventory/cloudforms.py]: Oh, I see you've made the `actions` configurable.  Disregard that comment.

**Comment by mrjoshuap** [comment]
flake8 returns all clear with exception of long lines, some of which I'm not sure the best way to resolve.  Are we extremely picky on that front?

**Comment by mrjoshuap** [comment]
For some reason, I removed the purge_actions method and forgot to add it back in.  I fixed that.  Also, added a nest_tags option so that CloudForms tags would show up as a nested list.  I split them based on the / in the tag name.  It is disabled by default because my sample size is vary small and there might be problems when a node with the same name appears in multiple tags.

**Comment by mrjoshuap** [comment]
Here's some sample outputs....

[cloudforms_py--host.json.txt](https://github.com/ansible/ansible/files/425403/cloudforms_py--host.json.txt)
[cloudforms_py--list--nested-tags.json.txt](https://github.com/ansible/ansible/files/425404/cloudforms_py--list--nested-tags.json.txt)
[cloudforms_py--list.json.txt](https://github.com/ansible/ansible/files/425405/cloudforms_py--list.json.txt)

**Comment by jlaska** [comment]
> flake8 returns all clear with exception of long lines, some of which I'm not sure the best way to resolve. Are we extremely picky on that front?

I'm not, I don't believe @nitzmahone is either.

File: contrib/inventory/cloudforms.py [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+                    self.inventory['tags'] = dict(children=[], vars={}, hosts=[])
+
+                if not self.cloudforms_nest_tags:
+                    # don't expand tags, just use them in a safe way
+                    for group in host['tags']:
+                        # Add sub-group, as a child of top-level
+                        safe_key = self.to_safe(group['name'])
+                        if safe_key:
+                            if self.args.debug:
+                                print "Adding sub-group '%s' to parent 'tags'" % safe_tag_name
```
Comment (jlaska) [inline contrib/inventory/cloudforms.py]: Heads  up, I think `safe_tag_name` is undefined here

Comment (mrjoshuap) [inline contrib/inventory/cloudforms.py]: you are correct sir, my bad, I missed the output from flake8!  /sigh

**Comment by nitzmahone** [comment]
Nope- we're trying to tighten up pep/flake kinds of things in the core codebase going forward, but this being contrib, we probably don't need to be too militant. :)

**Comment by jlaska** [comment]
@nitzmahone looks good to me, thanks @mrjoshuap for all the work!

**Pull request merged**